nanachan/extensions/profiles.py

```diff
@@ -119,229 +104,328 @@ async def promo(self, ctx: Context):
+    async def _photo_button_cb(self, interaction: Interaction):
+        await interaction.response.send_message('Upload your profile picture', ephemeral=True)
+
+        def check(message):
+            return all(
+                [
+                    message.command is None,
+                    message.author == interaction.user,
+                    message.channel == interaction.channel,
+                ]
+            )
+
+        resp = await MultiplexingContext.set_will_delete(check=check)
+        resp = resp.message
+        if len(resp.attachments) > 0:
+            attachment = resp.attachments[0]
+            if attachment.content_type == 'image/png':
+                hikari = await to_producer(attachment.url)
+                self.profile.photo = hikari['url']
+            else:
+                await resp.reply('Not a valid PNG file!')
+
+        await resp.delete()
+        await self._edit_embed(self.profile, interaction)
+
+    async def _n7_major_select_cb(self, interaction: Interaction):
+        await interaction.response.defer()
+        assert interaction.data
+        assert 'values' in interaction.data
+        self.profile.n7_major = interaction.data['values'][0]
+        await self._edit_embed(self.profile, interaction)
+
+    async def _confirm_button_cb(self, interaction: Interaction):
+        await interaction.response.defer()
+        profile_to_send = UpsertProfileBody(
+            discord_username=self.member.name,
+            birthday=self.profile.birthday,
+            full_name=self.profile.full_name,
+            graduation_year=self.profile.graduation_year,
+            n7_major=self.profile.n7_major,
+            photo=self.profile.photo,
+            pronouns=self.profile.pronouns,
+            telephone=self.profile.telephone,
+        )
+        await Profiles._create_or_update_profile(self.member, profile_to_send)
+        assert interaction.message
+        await interaction.message.edit(
+            content='Profile has been updated successfully.', embed=None, view=None
+        )
+
+    async def _cancel_button_cb(self, interaction: Interaction):
+        assert interaction.message
+        await interaction.message.edit(content='Profile update cancelled.', embed=None, view=None)
+
+    async def _instantiate_form_modal(self, interaction: Interaction, discord_id: int):
+        modal = ProfileModal(title='Create/Update your Japan7 profile.', profile=self.profile)
+        await interaction.response.send_modal(modal)
+        await modal.wait()
+        self._update_profile_from_dict(self.profile, modal.profile_dict)
+        await self._edit_embed(self.profile, interaction)
+
+    async def interaction_check(self, interaction: Interaction):
+        return self.member.id == interaction.user.id
+
+    @staticmethod
+    def _update_profile_from_dict(profile: ProfileSearchResult, profile_dict: dict[str, str]):
+        for key, val in profile_dict.items():
+            if hasattr(profile, key):
```

Comment:
doesn’t seem necessary, profile_dict could be a TypedDict to ensure we don’t add keys we don’t intend to use.